src/Core/Optimise/Reduce.hs

```diff
@@ -1,426 +1,273 @@
+    -> changed a'
+    | otherwise -> pure (App f' x')
+
+reduceTerm t@Match{} = reduceTermK t pure
+reduceTerm t@Let{}   = reduceTermK t pure
+
+-- | Reduce the provided term, running the continuation when reaching the
+-- "leaf" node.
+--
+-- This allows us to implement commuting conversion for lets and matches
+-- in a more intuitive manner.
+reduceTermK :: forall a m. MonadReduce a m
+            => Term a
+            -> (Term a -> m (Term a))
+            -> m (Term a)
+
+reduceTermK (Let (One (v, ty, e)) rest) cont = reduceTermK e $ \e' -> do
+  s <- ask
+  case e' of
+    -- Let of bottom conversion: we've errored here, so we can skip any
+    -- remaining code.
+    App (Ref f _) msg
+      | toVar (lookupRawVar f s) == vError
+      ->
+        let Just ty = approximateType rest -- TODO: Is this valid with our use of cont?
+            errTy = ForallTy Irrelevant (tyString :: Type a)
+            na = fromVar tyvarA :: a
+        in changed $
+          Let (One ( v, errTy ty
+                   , TyApp (Ref (fromVar vError) (ForallTy (Relevant na) StarTy (errTy (VarTy na)))) ty))
+            (App (Ref v (errTy ty)) msg)
+
+    _ -> do
+      rest' <- local (extendVar (v, ty, e')) (reduceTermK rest cont)
+      s <- ask
+      if
+        -- If we're binding a trivial atom, then we can strip it - we'll have
+        -- inlined it elsewhere and so it's dead.
+        | Atom a <- e', trivialAtom a -> changed rest'
+
+        -- Eta conversion for simple lets
+        | Atom (Ref v' _) <- rest', v == v' -> changed e'
+
+        -- Eta conversion for single constructor types
+        | Atom (Lit Unit) <- rest', ty == tyUnit -> changed e'
+        | Atom (Ref _ ty') <- rest'
+        , ty `unifyClosed` ty'
+        , Just tyName <- unwrapTy ty
+        , Just [_] <- VarMap.lookup (toVar tyName) (s ^. types)
+        -> changed e'
+
+        -- Match commuting conversion for multiple arms
+        | Match test arms <- e'
+        , Just restTy <- approximateType rest'
+        -> do
+            join <- fresh' ValueVar
+            let joinTy = ForallTy Irrelevant ty restTy
+                joinVar = Ref join joinTy
+
+                shoveJoinArm :: Arm a -> m (Arm a) = armBody %%~ shoveJoin
+                shoveJoin (Let bind body) = Let bind <$> shoveJoin body
+                shoveJoin (Match t bs) = Match t <$> traverse shoveJoinArm bs
+                shoveJoin (Atom a) = pure (App joinVar a)
+                shoveJoin ex = do
+                  var <- fresh' ValueVar
+                  pure (Let (One (var, ty, ex)) (App joinVar (Ref var ty)))
+
+            arms' <- traverse shoveJoinArm arms
+            changed $ Let (One (join, joinTy, Lam (TermArgument v ty) rest')) (Match test arms')
+
+        | otherwise -> pure (Let (One (v, ty, e')) rest')
+
+reduceTermK (Let (Many vs) rest) cont = do
+  -- We really can't do a lot with recursive lets sadly, so let's just
+  -- leave them for now.
+  vs' <- traverse (third3A reduceTerm) vs
```

Comment:
Will this η-reduce?

Reply:
#98 is still present - that will be fixed when loop breakers are added.